### PR TITLE
X11 windows implementation

### PR DESCRIPTION
**Pull request by Nihon13**

### CODE REVIEW

File: 3rdparty/CMakeLists.txt [hunk 1]
```diff
@@ -542,6 +541,11 @@ if (NBL_COMPILE_WITH_SDL2 )
 	    SDL2-static
 	)
 endif()
+if(ANDROID AND NOT ${CMAKE_HOST_SYSTEM_NAME} STREQUAL "Windows")
+	list(APPEND NBL_3RDPARTY_TARGETS
+		OpenSSL
+	)
+endif()
```
Comment (devshgraphicsprogramming) [inline 3rdparty/CMakeLists.txt:548]: TODO: @AnastaZIuk ASAP, also make OpenSSL do a fully-out-of-source build without soiling the submodule

File: CMakeLists.txt [hunk 2]
```diff
@@ -96,7 +97,11 @@ if(NBL_BUILD_ANDROID)
 	endif()
 	
 	string(LENGTH ${Java_JAVA_EXECUTABLE} Java_JAVA_EXECUTABLE_LENGTH)
-	set(Java_JAVA_EXE_BIN_LENGTH 13)
+	if(MSVC)
+		set(Java_JAVA_EXE_BIN_LENGTH 13)
+	else()
+		set(Java_JAVA_EXE_BIN_LENGTH 9)
```
Comment (devshgraphicsprogramming) [inline CMakeLists.txt:103]: huh?

File: CMakeLists.txt [hunk 3]
```diff
@@ -412,7 +417,7 @@ elseif(ANDROID)
 	set(_NBL_PLATFORM_ANDROID_ 1)
 elseif(UNIX)
 	set(_NBL_PLATFORM_LINUX_ 1)
-	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -msse4.1")
+	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -msse4.1 -fpermissive")
```
Comment (devshgraphicsprogramming) [inline CMakeLists.txt:420]: @Nihon13 what did you need to enable permissive for?

File: examples_tests/52.SystemTest/main.cpp [hunk 4]
```diff
@@ -205,11 +202,12 @@ class WindowEventCallback : public IWindow::IEventCallback
...
 	core::smart_refctd_ptr<InputSystem> m_inputSystem;
 	system::logger_opt_smart_ptr m_logger;
-	bool m_gotWindowClosedMsg;
 };
 
-int main()
+int main(int argc, char** argv)
 {
+	path CWD = path(argv[0]).parent_path().generic_string() + "/";
+	path mediaWD = CWD.generic_string() + "../../media/";
```
Comment (devshgraphicsprogramming) [inline examples_tests/52.SystemTest/main.cpp:210]: CommonAPI

File: examples_tests/52.SystemTest/main.cpp [hunk 5]
```diff
@@ -205,11 +202,12 @@ class WindowEventCallback : public IWindow::IEventCallback
 private:
 	core::smart_refctd_ptr<InputSystem> m_inputSystem;
 	system::logger_opt_smart_ptr m_logger;
-	bool m_gotWindowClosedMsg;
 };
 
-int main()
+int main(int argc, char** argv)
 {
+	path CWD = path(argv[0]).parent_path().generic_string() + "/";
```
Comment (devshgraphicsprogramming) [inline examples_tests/52.SystemTest/main.cpp:209]: @sadiuk some sort of a static in `system` namespace which is less hurtful to the eyes

File: src/nbl/CMakeLists.txt [hunk 6]
```diff
@@ -629,7 +629,7 @@ if (_NBL_COMPILE_WITH_OPEN_EXR_)
 endif()
 nbl_install_lib(png_static)
 
-if(NOT (ANDROID AND "${CMAKE_HOST_SYSTEM_NAME}" STREQUAL "Windows")) # OpenSSL can't be built for Android platform using windows OS host
+if(ANDROID AND NOT "${CMAKE_HOST_SYSTEM_NAME}" STREQUAL "Windows") # OpenSSL can't be built for Android platform using windows OS host
```
Comment (devshgraphicsprogramming) [inline src/nbl/CMakeLists.txt:632]: @AnastaZIuk to fix soon

File: CMakeLists.txt (hunk 2 above)
Comment (Nihon13) [reply CMakeLists.txt:103]: That was borrowed from Danylo ("Some gcc fixes + Arek's cmake fixes" commit).

File: CMakeLists.txt (hunk 3 above)
Comment (Nihon13) [reply CMakeLists.txt:420]: Without -fpermissive flag, i couldn't build `CDirQuantCacheBase.h` file.
GCC error: there are no arguments to ‘getSerializedCacheSizeInBytes’ that depend on a template parameter, so a declaration of ‘getSerializedCacheSizeInBytes’ must be available [-fpermissive]

Comment (devshgraphicsprogramming) [reply CMakeLists.txt:420]: that flag needs to get removed and that compile error fixed, maybe @Crisspl can help here.

File: CMakeLists.txt (hunk 2 above)
Comment (devshgraphicsprogramming) [reply CMakeLists.txt:103]: @sadiuk || @AnastaZIuk explain?

File: CMakeLists.txt (hunk 3 above)
Comment (Crisspl) [reply CMakeLists.txt:420]: so just move definition/forward decl of getSerializedCacheSizeInBytes over its usages. If this doesnt help, id have to see whole exact build log

File: CMakeLists.txt (hunk 2 above)
Comment (devshgraphicsprogramming) [reply CMakeLists.txt:103]: ok @AnastaZIuk explain please?

Comment (AnastaZIuk) [reply CMakeLists.txt:103]: on linux there is no `.exe` extension on executables, so we need to remove 4 characters

Comment (devshgraphicsprogramming) [reply CMakeLists.txt:103]: ah ok

File: CMakeLists.txt (hunk 3 above)
Comment (Nihon13) [reply CMakeLists.txt]: Fixed.

File: include/nbl/ui/CWindowX11.h [hunk 7]
```diff
@@ -1,14 +1,13 @@
...
-#ifdef _NBL_PLATFORM_LINUX_x1
+#ifdef _NBL_PLATFORM_LINUX_
 #include "nbl/ui/IWindowX11.h"
 
-
 #include <X11/Xutil.h>
 #include <X11/extensions/xf86vmode.h>
 #include <X11/extensions/Xrandr.h>
-#include "nbl_os.h"
+// #include "nbl_os.h" // @Nihon - no such file
```
Comment (devshgraphicsprogramming) [inline include/nbl/ui/CWindowX11.h:10]: there's `nbl/system/system.h` or something like that

File: include/nbl/ui/CWindowManagerX11.h [hunk 8]
```diff
@@ -69,6 +70,38 @@ static Xinput xinput("Xinput");
...
+
+        std::map<Key, Val> m_map;
+
+        inline void insert(const Key& _key, const Val& _val)
+        {
+            auto lk = lock_write();
+            m_map.insert(std::make_pair(_key, _val));
+        }
+
+        inline Val& read(const Key& _object)
```
Comment (devshgraphicsprogramming) [inline include/nbl/ui/CWindowManagerX11.h]: call it `get`

Comment (devshgraphicsprogramming) [reply include/nbl/ui/CWindowManagerX11.h]: or `find`

Comment (devshgraphicsprogramming) [reply include/nbl/ui/CWindowManagerX11.h]: use the `m_map.find` method and handle the case where object is not found `found==end()`, if the function MUST always find the object (by logical construction) then `assert`

File: include/nbl/ui/CWindowManagerX11.h [hunk 9]
```diff
@@ -21,6 +21,8 @@ NBL_SYSTEM_DECLARE_DYNAMIC_FUNCTION_CALLER_CLASS(X11, system::DefaultFuncPtrLoad
     ,XGetVisualInfo
     ,XCreateColormap
     ,XCreateWindow
+    ,XDestroyWindow
+    ,XCreateSimpleWindow
```
Comment (devshgraphicsprogramming) [inline include/nbl/ui/CWindowManagerX11.h:25]: whats the difference between `XCreateWindow` and `XCreateSimpleWindow`

File: src/nbl/ui/CWindowX11.cpp [hunk 10]
```diff
@@ -205,25 +191,43 @@ core::vector<XID> CWindowManagerX11::getConnectedKeyboards() const
...
 
 CWindowX11::~CWindowX11()
 {
-    m_manager->destroyWindow(this);
+    x11.pXDestroyWindow(m_dpy, m_native);
+}
+
+void CWindowManagerX11::CThreadHandler::init()
+{
+    m_dpy = x11.pXOpenDisplay((char *)0);
```
Comment (devshgraphicsprogramming) [inline src/nbl/ui/CWindowX11.cpp:210]: why open a display twice, is that even allowed?

File: include/nbl/ui/CWindowManagerX11.h (hunk 9 above)
Comment (Nihon13) [reply include/nbl/ui/CWindowManagerX11.h:25]: In `XCreateWindow` you can set more window's attributes.
https://tronche.com/gui/x/xlib/window/XCreateWindow.html

Comment (devshgraphicsprogramming) [reply include/nbl/ui/CWindowManagerX11.h:25]: ok doesn't seem like non-simple lets us set anything useful.

File: src/nbl/ui/CWindowX11.cpp [hunk 11]
```diff
@@ -36,92 +36,90 @@ CWindowManagerX11::CWindowManagerX11()
...
+                {
                     x11.pXMoveWindow(m_dpy, m_native, e.x, e.y);
+                }
+            }
+            break;
+        }
+
+        case MapNotify: // window show event
+        {
+            std::cout << "Show event!\n";
```
Comment (devshgraphicsprogramming) [inline src/nbl/ui/CWindowX11.cpp:74]: the example 52 (or some similar number) already has all the logging for these window events, just use the callback right

Comment (Nihon13) [reply src/nbl/ui/CWindowX11.cpp:74]: Yea, i forgot to delete it. It was just for testing.

**Pull request closed without merging**